src/core/protocol-authorization.ts

```diff
@@ -333,17 +328,42 @@ export class ProtocolAuthorization {
     }
   }
 
+  private static async getInboundMessageActions(
+    tenant: string,
+    incomingMessage: RecordsRead | RecordsWrite,
+    messageStore: MessageStore,
+  ): Promise<ProtocolAction[]> {
+    if (incomingMessage.message.descriptor.method === DwnMethodName.Read) {
+      return [ProtocolAction.Read];
+    }
+    // else 'Write'
+
+    const incomingRecordsWrite = incomingMessage as RecordsWrite;
+    if (await incomingRecordsWrite.isInitialWrite()) {
+      // only 'write' allows initial RecordsWrites; 'update' only applies to subsequent RecordsWrites
+      return [ProtocolAction.Write];
+    } else if (await incomingRecordsWrite.authoredByInitialRecordAuthor(tenant, messageStore)) {
+      // Both 'update' and 'write' authorize the incoming message
+      return [ProtocolAction.Write, ProtocolAction.Update];
+    } else {
+      // Actors other than the initial record author must be authorized to 'update' the message
+      return [ProtocolAction.Update];
+    }
+  }
+
   /**
    * Verifies the action (e.g. read/write) specified in the given message matches the allowed actions in the rule set.
    * @throws {Error} if action not allowed.
    */
   private static async verifyAllowedActions(
+    tenant: string,
     incomingMessage: RecordsRead | RecordsWrite,
     inboundMessageRuleSet: ProtocolRuleSet,
     ancestorMessageChain: RecordsWriteMessage[],
+    messageStore: MessageStore,
   ): Promise<void> {
     const incomingMessageMethod = incomingMessage.message.descriptor.method;
-    const inboundMessageAction = methodToAllowedActionMap[incomingMessageMethod];
+    const inboundMessageActions = await ProtocolAuthorization.getInboundMessageActions(tenant, incomingMessage, messageStore);
```

Comment:
Naming is a bit vague to me, consider renaming to perhaps `getActionsSeekingARuleMatch()` or `getPotentiallyAllowedActions()` or `getActionCandidates()`

```suggestion
    const actionsSeekingARuleMatch = await ProtocolAuthorization.getActionsSeekingARuleMatch(tenant, incomingMessage, messageStore);
```

Reply:
Updated